frontend/src/InputForm.js

```diff
@@ -0,0 +1,791 @@
+    this.updateRankSelectOptions = this.updateRankSelectOptions.bind(this);
+    
+    // lecture sections
+    this.updateLectureSectionProfessor = this.updateLectureSectionProfessor.bind(this);
+    this.updateLectureSectionStartTime = this.updateLectureSectionStartTime.bind(this);
+    this.updateLectureSectionEndTime = this.updateLectureSectionEndTime.bind(this);
+    this.updateLectureSectionDays = this.updateLectureSectionDays.bind(this);
+
+    // lab sections
+    this.updateLabSectionProfessor = this.updateLabSectionProfessor.bind(this);
+    this.updateLabSectionStartTime = this.updateLabSectionStartTime.bind(this);
+    this.updateLabSectionEndTime = this.updateLabSectionEndTime.bind(this);
+    this.updateLabSectionDays = this.updateLabSectionDays.bind(this);
+
+    // criterion
+    this.updateTimeStartPreference = this.updateTimeStartPreference.bind(this);
+    this.updateTimeEndPreference = this.updateTimeEndPreference.bind(this);
+    this.updateSubjectPreference = this.updateSubjectPreference.bind(this);
+    
+    // basic info
+    this.updateMinCredits = this.updateMinCredits.bind(this);
+    this.updateMaxCredits = this.updateMaxCredits.bind(this);
+    this.updateTermStartDate = this.updateTermStartDate.bind(this);
+    this.updateTermEndDate = this.updateTermEndDate.bind(this);
+
+    this.convertCourseLectureSections = this.convertCourseLectureSections.bind(this);
+    this.convertCourseLabSections = this.convertCourseLabSections.bind(this);
+    this.submit = this.submit.bind(this);
+  }
+
+  // used to keep track of which ranks have been selected, so no two courses have the same rank 
+  selectedRanks = [];
+
+  updateCourseName(id, courseName) {
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.map((course, index) =>
+        index === id
+          ? ({ ...course, name: courseName })
+          : course)
+    });
+  }
+
+  updateCourseID(id, courseID) {
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.map((course, index) =>
+        index === id
+          ? ({ ...course, courseID: courseID })
+          : course)
+    });
+  }
+
+  updateCourseSubject(id, courseSubject) {
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.map((course, index) =>
+        index === id
+          ? ({ ...course, subject: courseSubject })
+          : course)
+    });
+  }
+
+  updateCourseCredits(id, courseCredits) {
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.map((course, index) =>
+        index === id
+          ? ({ ...course, credits: courseCredits })
+          : course)
+    });
+  }
+
+  updateCourseIsRequired(id) {
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.map((course, index) =>
+        index === id
+          ? ({ ...course, isRequired: !this.state.courses[index].isRequired })
+          : course)
+    });
+  }
+
+  updateCourseRank(id, selected) {
```

Comment:
(Didn't change nested ternary operators in udpateCourseRank() because they'll go away with the new scoring changes in the other branch, grandparentComponents)